File: bot/src/command/SettingsCommand.ts
import Command from "./Command";
import {Context} from "../Context";
import {VkMessage} from "../service/VkMessagesService";

export default class SettingsCommand extends Command {
    constructor(context: Context) {
        super(context);
    }

    getCommandShortUsage(): string {
        return '/sweet settings (...)';
    }

    canYouHandleThisCommand(command: string, message: VkMessage): boolean {
        return command === 'settings';
    }

    requiresPrivileges(peerId: number): boolean {
        return true;
    }

    // TODO refactor
    async handle(command: string, rawArguments: string, message: VkMessage): Promise<void> {
        const { vkMessagesService, chatSettingsService } = this.context;

        const subCommand = rawArguments.split(" ")[0];
        const text = rawArguments.slice(subCommand.length).trim();
        const settings = await chatSettingsService.getSettingsOrCreateDefault(message.peerId);

        if (!subCommand) {
            let response = ``;
            response += `max_output_tokens=${settings.gptMaxOutputTokens}\n`;
            response += `max_input_tokens=${settings.gptMaxInputTokens}\n`;
            response += `temperature=${settings.gptTemperature}\n`;
            response += `top_p=${settings.gptTopP}\n`;
            response += `frequency_penalty=${settings.gptFrequencyPenalty}\n`;
            response += `presence_penalty=${settings.gptPresencePenalty}\n`;
            response += `process_audio_messages=${settings.processAudioMessages}\n`;
            await vkMessagesService.send(message.peerId, response);
        } else if (subCommand == "get") {
            if (text.length == 0)
                await  vkMessagesService.send(message.peerId, this.getUsage());
            const settingName = text;
            let value = null;
            if (settingName == "max_output_tokens")
                value = settings.gptMaxOutputTokens;
            if (settingName == "max_input_tokens")
                value = settings.gptMaxInputTokens;
            if (settingName == "temperature")
                value = settings.gptTemperature;
            if (settingName == "top_p")
                value = settings.gptTopP;
            if (settingName == "frequency_penalty")
                value = settings.gptFrequencyPenalty;
            if (settingName == "presence_penalty")
                value = settings.gptPresencePenalty;
            if (settingName == "process_audio_messages")
                value = settings.processAudioMessages;

            await vkMessagesService.send(message.peerId, `${settingName}=${value}`);
        } else if (subCommand == "set") {
            if (text.length == 0)
                await vkMessagesService.send(message.peerId, this.getUsage());
            let args = text.split("=").map(x => x.trim());
            if (args.length == 1)
                args = text.split(" ").filter(x => x != "").map(x => x.trim());
            if (args.length != 2)
                await vkMessagesService.send(message.peerId, this.getUsage());

            const settingName = args[0];
            const settingValue = args[1];
            if (settingName == "max_output_tokens") {
                const value = parseInt(settingValue);
                if (isNaN(value) || value < 1 || value > 2048) {
                    await vkMessagesService.send(message.peerId, `Это должно быть целое число от 1 до 2048`);
                    return;
                } else {
                    await chatSettingsService.setGptMaxOutputTokens(message.peerId, value);
                }
            } else if (settingName == "max_input_tokens") {
                const value = parseInt(settingValue);
                if (isNaN(value) || value < 0) {
                    await vkMessagesService.send(message.peerId, `Это должно быть положительное число`);
                    return;
                } else
                    await chatSettingsService.setGptMaxInputTokens(message.peerId, value);
            } else if (settingName == "temperature") {
                const value = parseFloat(settingValue);
                if (isNaN(value) || value < 0 || value > 2) {
                    await vkMessagesService.send(message.peerId, `Это должно быть число от 0 до 2`);
                    return;
                } else
                    await chatSettingsService.setGptTemperature(message.peerId, value);
            } else if (settingName == "top_p") {
                const value = parseFloat(settingValue);
                if (isNaN(value) || value < 0 || value > 1) {
                    await vkMessagesService.send(message.peerId, `Это должно быть число от 0 до 1`);
                    return;
                } else
                    await chatSettingsService.setGptTopP(message.peerId, value);
            } else if (settingName == "frequency_penalty") {
                const value = parseFloat(settingValue);
                if (isNaN(value) || value < 0 || value > 2) {
                    await vkMessagesService.send(message.peerId, `Это должно быть число от 0 до 2`);
                    return;
                } else
                    await chatSettingsService.setGptFrequencyPenalty(message.peerId, value);
            } else if (settingName == "presence_penalty") {
                const value = parseFloat(settingValue);
                if (isNaN(value) || value < 0 || value > 2) {
                    await vkMessagesService.send(message.peerId, `Это должно быть число от 0 до 2`);
                    return;
                } else {
                    await chatSettingsService.setGptPresencePenalty(message.peerId, value);
                }
            } else if (settingName == "process_audio_messages") {
                const value = settingValue == "true" || settingValue == "yes";
                await chatSettingsService.setProcessAudioMessages(message.peerId, value);
            } else {
                await vkMessagesService.send(message.peerId, `Нет такого параметра`);
            }

            await vkMessagesService.send(message.peerId, `${settingName}=${settingValue}`);
        } else {
            await vkMessagesService.send(message.peerId, this.getUsage());
        }
    }

    private getUsage(): string {
        let usage = `Команды:\n`;
        usage += `/sweet settings\n`;
        usage += `/sweet settings help\n`;
        usage += `/sweet settings get (имя)\n`;
        usage += `/sweet settings set (имя)[=](значение)\n`;
        return usage;
    }
}

File: backend/src/main/java/space/davids_digital/sweetie/command/SettingsCommand.kt
package space.davids_digital.sweetie.command

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.springframework.stereotype.Component
import space.davids_digital.sweetie.integration.vk.VkMessageService
import space.davids_digital.sweetie.model.VkMessageModel
import space.davids_digital.sweetie.service.ChatSettingsService

@Component
class SettingsCommand(
    private val vkMessageService: VkMessageService,
    private val chatSettingsService: ChatSettingsService
): Command {
    override fun getNames(): Array<String> {
        return arrayOf("settings")
    }

    override fun getUsage(): String {
        return "settings (...)"
    }

    override fun requiresChatAdmin(): Boolean {
        return true
    }

    override fun requiresAppCeo(): Boolean {
        return false
    }

    override suspend fun handle(commandName: String, rawArguments: String, message: VkMessageModel) {
        val subCommand = rawArguments.split(" ")[0]
        val rest = rawArguments.substring(subCommand.length).trim()
        when (subCommand) {
            "" -> handleDefault(message)
            "set" -> handleSet(message, rest)
            else -> handleHelp(message)
        }
    }

    private fun handleDefault(message: VkMessageModel) {
        val settings = chatSettingsService.getOrCreateDefault(message.peerId)
        with(settings) {
            vkMessageService.send(message.peerId, """
                max_output_tokens=$gptMaxOutputTokens
                max_input_tokens=$gptMaxInputTokens
                temperature=$gptTemperature
                top_p=$gptTopP
                frequency_penalty=$gptFrequencyPenalty
                presence_penalty=$gptPresencePenalty
                process_audio_messages=$processAudioMessages
                tts_voice=$ttsVoice
                tts_speed=$ttsSpeed
            """.trimIndent())
        }
    }

    private suspend fun handleSet(message: VkMessageModel, rest: String) {
        if (rest.isBlank()) {
            handleHelp(message)
            return
        }
        var args = rest.split("=").map { it.trim() }
        if (args.size == 1) {
            args = rest.split(" ").map { it.trim() }.filter { it.isNotBlank() }
        }
        if (args.size != 2) {
            handleHelp(message)
            return
        }
        val settingName = args[0]
        val settingValue = args[1]
        var settings = withContext(Dispatchers.IO) {
            chatSettingsService.getOrCreateDefault(message.peerId)
        }
        try {
            settings = when (settingName) {
                "max_output_tokens" -> settings.copy(gptMaxOutputTokens = settingValue.toInt())
                "max_input_tokens" -> settings.copy(gptMaxInputTokens = settingValue.toInt())
                "temperature" -> settings.copy(gptTemperature = settingValue.toDouble())
                "top_p" -> settings.copy(gptTopP = settingValue.toDouble())
                "frequency_penalty" -> settings.copy(gptFrequencyPenalty = settingValue.toDouble())
                "presence_penalty" -> settings.copy(gptPresencePenalty = settingValue.toDouble())
                "process_audio_messages" -> settings.copy(processAudioMessages = settingValue.toBoolean())
                "tts_voice" -> settings.copy(ttsVoice = settingValue)
                "tts_speed" -> settings.copy(ttsSpeed = settingValue.toDouble())
                else -> {
                    vkMessageService.send(message.peerId, "Нет такой настройки")
                    return
                }
            }
        } catch (e: NumberFormatException) {
            vkMessageService.send(message.peerId, "Надо число")
            return
        }
        if (settings.gptMaxInputTokens < 0 || settings.gptMaxInputTokens > 16384) {
            vkMessageService.send(message.peerId, "Надо число от 0 до 16384")
            return
        }
        if (settings.gptMaxOutputTokens < 1 || settings.gptMaxOutputTokens > 2048) {
            vkMessageService.send(message.peerId, "Надо число от 1 до 2048")
            return
        }
        if (settings.gptTemperature < 0 || settings.gptTemperature > 2) {
            vkMessageService.send(message.peerId, "Надо число от 0 до 2")
            return
        }
        if (settings.gptTopP < 0 || settings.gptTopP > 1) {
            vkMessageService.send(message.peerId, "Надо число от 0 до 1")
            return
        }
        if (settings.gptFrequencyPenalty < 0 || settings.gptFrequencyPenalty > 2) {
            vkMessageService.send(message.peerId, "Надо число от 0 до 2")
            return
        }
        if (settings.gptPresencePenalty < 0 || settings.gptPresencePenalty > 2) {
            vkMessageService.send(message.peerId, "Надо число от 0 до 2")
            return
        }
        if (settings.ttsSpeed < 0.25 || settings.ttsSpeed > 4) {
            vkMessageService.send(message.peerId, "Надо число от 0.25 до 4")
            return
        }
        withContext(Dispatchers.IO) {
            chatSettingsService.updateSettings(message.peerId) {
                settings
            }
        }
        vkMessageService.send(message.peerId, "Ок")
    }

    private fun handleHelp(message: VkMessageModel) {
        vkMessageService.send(message.peerId, """
            Команды:
            /sweet settings
            /sweet settings help
            /sweet settings set (имя)[=](значение)
        """.trimIndent())
    }
}